Update `traits/resolution.md` (#1494)
rust-lang/rustc-dev-guide: src/traits/resolution.md

### Candidate assembly

> **TODO**: Talk about _why_ we have different candidates, and why it needs to happen in ap robe.

Author: ```suggestion
**TODO**: Talk about _why_ we have different candidates, and why it needs to happen in a probe.
```

The subroutines that decide whether a particular impl/where-clause/etc applies
to a particular obligation are collectively referred to as the process of

> _matching_. As of <!-- date-check --> Oct 2022, For `impl` candidates,

Reviewer: This can also be written like ...

```suggestion
_matching_. For `impl` candidates <!-- date-check: Oct 2022 -->,
```

Note that the candidate impl is chosen based on the `Self` type, but
confirmation is done based on (in this case) the `Target` type parameter.

### Selection during codegen

As mentioned above, during type checking, we do not store the results of trait
selection. At codegen time, we repeat the trait selection to choose a particular

> impl for each method call. This is done using `fn codegen_select_candidate`. In this second selection, we do not consider any
> where-clauses to be in scope because we know that each resolution will resolve
> to a particular impl.

Author: ```suggestion
impl for each method call. This is done using `fn codegen_select_candidate`. 
In this second selection, we do not consider any where-clauses to be in scope 
because we know that each resolution will resolve to a particular impl.
```

impl for each method call. This is done using `fn codegen_select_candidate`. In this second selection, we do not consider any
where-clauses to be in scope because we know that each resolution will resolve
to a particular impl.

One interesting twist has to do with nested obligations. In general, in trans,
we only need to do a "shallow" selection for an obligation. That is, we wish to
identify which impl applies, but we do not (yet) need to decide how to select
any nested obligations. Nonetheless, we *do* currently do a complete resolution,
and that is because it can sometimes inform the results of type inference.
One interesting twist has to do with nested obligations. In general, in codegen,
we only to figure out which candidate applies, we do not care about nested obligations, as these are already assumed to be true. Nonetheless, we *do* currently do fulfill all of them. That is because it can sometimes inform the results of type inference.
That is, we do not have the full substitutions in terms of the type variables
of the impl available to us, so we must run trait selection to figure
everything out.

> **TODO**: is this still talking about trans?

Author: We removed this following section because it's not yet implemented, is kinda an implementation detail, and the example is not exactly where the consideration above ("because it can sometimes inform the results of type inference") actually matters, which is with projection types specifically.
